### PR TITLE
[workflow] Workflow queue

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

Implementation of https://github.com/ray-project/ray/issues/24029

Here are the behaviors this PR supports and tests:

- [x] The queued workflow has a `PENDING` status
- [x] When a running workflow completes, a pending workflow from the queue starts running.
- [x] If the workflow queue is full, submitting a workflow raises `queue.Full("Workflow queue has been full")` immediately.
- [x] Submitting a workflow would not be blocked when the workflow is queued (pended).
- [x] `workflow.run()` or `ray.get(workflow.async_run())` would block on the pending workflow, until the workflow resumes running and finishes.
- [x] `ray.workflow.get_output_async(workflow_id)` would not be blocked when the workflow is queued (pended).
- [x] `ray.workflow.get_output(workflow_id)` would block on the pending workflow, until the workflow resumes running and finishes.
- [x] Both pending and running workflows could be resumed, if the system was down accidentally.
- [x] When both running and pending workflows are resumed with ``workflow.resume_all()``, running workflows have the higher priority (i.e. the pending workflows would still likely be pending).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by suquark** [comment]
@iycheng @ericl This would be the workflow queue API. See if you like it.

The API requires users to specify `max_concurrent_jobs` and `max_queueing_jobs` in `workflow.init`. Once specified, these options would be permanent, you cannot change them unless tearing down the whole Ray cluster.

File: python/ray/workflow/api.py [hunk 1]
```diff
@@ -41,7 +41,7 @@
 
 
 @PublicAPI(stability="beta")
-def init() -> None:
+def init(max_concurrent_jobs=-1, max_queueing_jobs=-1) -> None:
```
Comment (ericl) [inline python/ray/workflow/api.py:44]:
```suggestion
def init(max_running_workflows=-1, max_pending_workflows=-1) -> None:
```

**Review by ericl — COMMENTED** [review]
Do we need to add a new PENDING/QUEUED workflow state?

**Comment by suquark** [comment]
This PR is blocked by #24767 for workflow status updating (adding new status).

File: python/ray/workflow/api.py [hunk 2]
```diff
@@ -41,11 +41,17 @@
 
 
 @PublicAPI(stability="beta")
-def init() -> None:
+def init(*, max_running_workflows: int = -1, max_pending_workflows: int = -1) -> None:
```
Comment (iycheng) [inline python/ray/workflow/api.py:44]: should the default one be None?

I'm thinking about the use case where the cluster is up and the workflow actor is up. Then the user wants to submit the jobs. THen does this mean the user has to always specify the right number?

File: python/ray/workflow/workflow_access.py [hunk 3]
```diff
@@ -170,16 +195,35 @@ def run_or_resume(
...
+            and len(self._running_workflows) >= self._max_running_workflows
+        ):
+            try:
+                self._workflow_queue.put_nowait((job_id, workflow_id))
+                self._queued_workflows.add(workflow_id)
+            except queue.Full:
+                # override with our error message
+                raise queue.Full("Workflow queue has been full") from None
+        else:
+            self._run_or_resume(job_id, workflow_id)
```
Comment (iycheng) [inline python/ray/workflow/workflow_access.py:216]: How about the return type here. Can we still get the result with ray.get ? Is there some py tricks we can use here?

**Comment by iycheng** [comment]
Also, how about resuming all? It seems that the queue reconstructed doesn't take order into consideration.

**Comment by suquark** [comment]
too messy to merge. let me squash and rebase it. the comment are easy to locate in the code

**Comment by suquark** [comment]
@iycheng ready for review

File: python/ray/workflow/workflow_storage.py [hunk 4]
```diff
@@ -229,10 +229,13 @@ class WorkflowStorage:
...
 
-    def __init__(self, workflow_id: str):
+    def __init__(self, workflow_id: str, _no_init_workflow: bool = False):
         from ray.workflow.api import _ensure_workflow_initialized
 
-        _ensure_workflow_initialized()
+        if not _no_init_workflow:
+            # Avoid initialize workflow when it is not necessary
+            # or would result in abnormality
+            _ensure_workflow_initialized()
```
Comment (iycheng) [inline python/ray/workflow/workflow_storage.py:238]: Seems hacky to me. Should we just fix _ensure_workflow_initialized?

File: python/ray/workflow/workflow_access.py [hunk 5]
```diff
@@ -378,24 +484,44 @@ def report_success(self, workflow_id: str) -> None:
...
 
-def init_management_actor() -> None:
-    """Initialize WorkflowManagementActor"""
+def init_management_actor(
+    max_running_workflows: Optional[int], max_pending_workflows: Optional[int]
+) -> None:
+    """Initialize WorkflowManagementActor.
+
+    Args:
+        max_running_workflows: The maximum number of concurrently running workflows.
```
Comment (iycheng) [inline python/ray/workflow/workflow_access.py:499]: Add documentation about what does None mean.

File: python/ray/workflow/workflow_access.py [hunk 6]
```diff
@@ -121,14 +123,40 @@ class LatestWorkflowOutput:
...
+
+# TODO(suquark): what should be the default concurrency limit?
+@ray.remote(num_cpus=0)
+class WorkflowResultRespondingActor:
+    """This async actor is for responding pending workflow results."""
+
+    def __init__(self):
+        self.running_or_pending_workflows: Dict[str, asyncio.Future] = {}
+
+    def set_running_workflow_output(self, workflow_id: str, output: WorkflowStaticRef):
```
Comment (iycheng) [inline python/ray/workflow/workflow_access.py:141]: I'm not sure, but should this function be async as well?

File: python/ray/workflow/workflow_access.py [hunk 7]
```diff
@@ -121,14 +123,40 @@ class LatestWorkflowOutput:
...
+        if workflow_id not in self.running_or_pending_workflows:
+            self.running_or_pending_workflows[workflow_id] = asyncio.Future()
+        self.running_or_pending_workflows[workflow_id].set_result(output)
+
+    def set_pending_workflow(self, workflow_id: str):
+        self.running_or_pending_workflows[workflow_id] = asyncio.Future()
```
Comment (iycheng) [inline python/ray/workflow/workflow_access.py:147]: Can we make this async and then get_output seems not necessary

```
async  def set_pending_workflow(self, workflow_id: str):
       self.running_or_pending_workflows[workflow_id] = asyncio.Future()
       return await self.running_or_pending_workflows[workflow_id]
```

File: python/ray/workflow/workflow_access.py [hunk 8]
```diff
@@ -121,14 +123,40 @@ class LatestWorkflowOutput:
...
+def _get_workflow_storage(workflow_id: str) -> workflow_storage.WorkflowStorage:
+    # Avoid initialize workflow in the workflow management actor,
+    # because it would invoke remote methods of itself,
+    # resulting in deadlock.
+    return workflow_storage.WorkflowStorage(workflow_id, _no_init_workflow=True)
+
+
+# TODO(suquark): what should be the default concurrency limit?
+@ray.remote(num_cpus=0)
+class WorkflowResultRespondingActor:
```
Comment (iycheng) [inline python/ray/workflow/workflow_access.py:135]: I feel the API here is a little bit confusing, for example, it'll set the default value if it doesn't exist. If this happens, does this mean there is a bug that we haven't caught?


Is it possible to have only two API for simplicity:
   register_workflow -> ObjectRef
   set_output -> ObjectRef
?

**Review by fishbone — CHANGES_REQUESTED** [review]
Overall it looks good. Please check the comments.

File: python/ray/workflow/workflow_access.py (hunk 6 above)
Comment (suquark) [reply python/ray/workflow/workflow_access.py:141]: Currently seem not necessary, as `await` is not used inside.

File: python/ray/workflow/workflow_access.py (hunk 7 above)
Comment (suquark) [reply python/ray/workflow/workflow_access.py:147]: hmmm, get output is used in another place for the `get_output` API. I still prefer to keep them seperated.

File: python/ray/workflow/workflow_access.py [hunk 9]
```diff
@@ -121,21 +123,76 @@ class LatestWorkflowOutput:
...
```
Comment (iycheng) [inline python/ray/workflow/workflow_access.py:135]: @yuanchi2807 @klwuibm

**Comment by suquark** [comment]
@iycheng This is a new implementation of workflow queue based on the new workflow refactoring. Ready for review.

**Comment by suquark** [comment]
Some tests will be affected by #26318, so I will retest them later.

File: doc/source/workflows/management.rst [hunk 10]
```diff
@@ -82,12 +83,22 @@ Storage Configuration
...
 
+Concurrency Control
+-------------------
+Ray workflow supports concurrency control. You can support the maximum running workflows and maximum pending workflows via ``workflow.init()``
+before executing any workflow. ``workflow.init()`` again with a different configuration would raise an error.
+
+For example, ``workflow.init(max_running_workflows=10, max_pending_workflows=50)`` means there will be at most 10 workflows running, 50 workflows pending.
+Submitting workflows when the number of pending workflows are at maximum would raise ``queue.Full("Workflow queue has been full")``. Getting the output of a pending workflow would be blocking until the workflow finishes running later.
+
+Pending workflows have the ``PENDING`` status, they can be resumed same as these running workflows if they failed with a system error.
```
Comment (iycheng) [inline doc/source/workflows/management.rst:100]: What does this mean?

File: python/ray/workflow/api.py [hunk 11]
```diff
@@ -33,21 +33,50 @@
...
+        max_pending_workflows: The maximum number of queued workflows.
+            Use -1 as infinity. 'None' means preserving previous setting or initialize
+            the setting with infinity.
     """
     usage_lib.record_library_usage("workflow")
 
+    if max_running_workflows is not None:
+        if not isinstance(max_running_workflows, int):
+            raise TypeError("'max_running_workflows' must be None or an integer.")
+        if max_running_workflows < -1 or max_running_workflows == 0:
```
Comment (iycheng) [inline python/ray/workflow/api.py:59]: Why not `max_running_workflows <= 0` ?

File: python/ray/workflow/tests/test_workflow_queuing.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
+import pytest
```
[the remark this answers is not in the document]
Comment (iycheng) [reply python/ray/workflow/tests/test_workflow_queuing.py:1]: Please address the comment

**Review by fishbone — COMMENTED** [review]
Overall it looks good and the new engine simplifies a lot of things. Could you please also address the comments I left in the previous reviews?

File: doc/source/workflows/management.rst (hunk 10 above)
Comment (suquark) [reply doc/source/workflows/management.rst:100]: I mean if a workflow status is `RUNNING` and the workflow failed, we can resume the workflow. Same as the `PENDING` status.

File: python/ray/workflow/workflow_access.py (hunk 8 above)
Comment (suquark) [reply python/ray/workflow/workflow_access.py]: this comment is outdated

File: python/ray/workflow/workflow_access.py (hunk 7 above)
Comment (suquark) [reply python/ray/workflow/workflow_access.py]: this comment is outdated

File: python/ray/workflow/workflow_access.py (hunk 9 above)
Comment (suquark) [reply python/ray/workflow/workflow_access.py]: this is outdated

File: python/ray/workflow/tests/test_workflow_queuing.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+        (tmp_path / str(x)).write_text(str(x))
+        with filelock.FileLock(lock_path):
+            return x
+
+    with filelock.FileLock(lock_path):
+        wf_1 = workflow.run_async(long_running.bind(1))
+        wf_2 = workflow.run_async(long_running.bind(2))
+        time.sleep(5)
+        assert (tmp_path / str(1)).exists()
+        assert not (tmp_path / str(2)).exists()
```
Comment (iycheng) [inline python/ray/workflow/tests/test_workflow_queuing.py:127]: Also check the status of w1 and w2

**Review by fishbone — APPROVED** [review]
Commented.

One question about resume: for pending one's, when they are added back, are they still in the same order as before?

I am ok with either right now. But please add comments/doc about this as well.

**Comment by suquark** [comment]
CI failures are not related. I'll merge this PR.

**Pull request merged**